common/src/main/java/io/netty/util/ResourceLeakDetector.java

```diff
@@ -411,13 +411,17 @@ private void record0(Object hint) {
                 Record newHead;
                 boolean dropped;
                 do {
+                    dropped = false;
                     if ((prevHead = oldHead = headUpdater.get(this)) == null) {
                         // already closed.
                         return;
                     }
-                    int numElements = oldHead.pos + 1;
-                    if (dropped = PlatformDependent.threadLocalRandom().nextInt(1 << numElements) >= TARGET_RECORDS) {
-                        prevHead = oldHead.next;
+                    final int numElements = oldHead.pos + 1;
+                    if (numElements >= TARGET_RECORDS) {
+                        final int backOffFactor = numElements - TARGET_RECORDS;
+                        if (dropped = PlatformDependent.threadLocalRandom().nextInt(1 << backOffFactor) != 0) {
```

Comment:
@carl-mastrangelo shouldn't you not update `dropped` it anymore once you set `dropped` to `true` ?

Reply:
dropped needs to be reset because if it loses the compareAndSet race, it may retry without dropping a record the second time around.

Reply:
It could be added to an else clause instead of being done unconditionally at the top.

Reply:
`backOffFactor` may overflow. if `backOffFactor == 31` this will result in an IllegalArgumentException. It maybe unlikely as `numElements` grows, but consider protecting for this case just incase we are unlucky.

Reply:
Done